### PR TITLE
License Manager: Adding license manager

### PR DESCRIPTION
**Pull request by abhilashraju**
License manager changes to enable license feature

Tested: Tested by executing following curl requests 1) /redfish/v1/Registries/License(Get)
2) /redfish/v1/Registries/License/License
3) /redfish/v1/LicenseService(Get)
4) /redfish/v1/LicenseService/Licenses

Change-Id: I289ce73e0bb5c5dc77e696c18ee51ec9df92de00

### CODE REVIEW

**Comment by gtmills** [comment]
Please use a different branch name than 1110. Merge conflicts are really hard to solve if the branch name is the same. 
license-service, 1110-license-service is a much better branch name

File: meson.build [hunk 1]
```diff
@@ -310,7 +311,7 @@ boost = dependency(
   modules: [
     'url',
     ],
-  version : '>=1.84.0',
+  version : '>=1.83.0',
```
Comment (gtmills) [inline meson.build:314]: Don't change this

File: redfish-core/lib/license_service.hpp [hunk 2]
```diff
@@ -0,0 +1,678 @@
...
+                    "UnavailableOffline";
+            }
+        }
+    });
+}
+
+inline void requestRoutesLicenseEntry(App& app)
+{
+    BMCWEB_ROUTE(app, "/redfish/v1/LicenseService/Licenses/<str>/")
+        .privileges({{"Login"}})
```
Comment (gtmills) [inline redfish-core/lib/license_service.hpp:670]: GetLicense here

File: redfish-core/lib/license_service.hpp [hunk 3]
```diff
@@ -0,0 +1,678 @@
...
+            "#LicenseCollection.LicenseCollection";
+        asyncResp->res.jsonValue["@odata.id"] =
+            "/redfish/v1/LicenseService/Licenses";
+        asyncResp->res.jsonValue["Name"] = "License Collection";
+
+        getLicenseEntryCollection(asyncResp);
+    });
+
+    BMCWEB_ROUTE(app, "/redfish/v1/LicenseService/Licenses")
+        .privileges({{"ConfigureManager"}})
```
Comment (gtmills) [inline redfish-core/lib/license_service.hpp:432]: PatchLicenseCollection

File: redfish-core/lib/license_service.hpp [hunk 4]
```diff
@@ -0,0 +1,678 @@
...
+    }
+    else if (status == "com.ibm.License.LicenseManager.Status.Activated")
+    {
+        BMCWEB_LOG_INFO("License Activated");
+        messages::licenseInstalled(asyncResp->res, licenseString);
+    }
+    else
+    {
+        auto ires = asyncResp->res.resultInt();
+        BMCWEB_LOG_INFO("License Activation Failed: {}", ires);
```
Comment (gtmills) [inline redfish-core/lib/license_service.hpp:302]: ERROR here

**Comment by raviteja-b** [comment]
looks like this https://github.com/ibm-openbmc/bmcweb/pull/873 was closed as part of bmcweb fork rebase

File: meson.build (hunk 1 above)
Comment (abhilashraju) [reply meson.build:314]: oh i forgot to change it back

File: redfish-core/lib/license_service.hpp (hunk 2 above)
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:670]: does it really add any value now. The name was same for all these time

File: redfish-core/lib/license_service.hpp (hunk 3 above)
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:432]: what you mean by PatchLicenseCollection?

File: redfish-core/lib/license_service.hpp (hunk 2 above)
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:670]: You are talking about function name right? If that is the case please tell me the convention we should follow. Quick check with recently approved commit show this name requestRoutesThermalMetrics(App& app)

File: redfish-core/lib/license_service.hpp (hunk 3 above)
Comment (gtmills) [reply redfish-core/lib/license_service.hpp:432]: https://github.com/ibm-openbmc/bmcweb/blob/1060/redfish-core/lib/ethernet.hpp#L2281

Comment (gtmills) [reply redfish-core/lib/license_service.hpp:432]: But crap you need https://gerrit.openbmc.org/c/openbmc/bmcweb/+/70854

File: redfish-core/lib/license_service.hpp (hunk 2 above)
Comment (gtmills) [reply redfish-core/lib/license_service.hpp:670]: I am talking about the privileges

File: redfish-core/lib/license_service.hpp [hunk 5]
```diff
@@ -0,0 +1,678 @@
...
+            BMCWEB_LOG_ERROR("DBUS response error: Unable to set "
+                             "the LicenseString property {}",
+                             ec);
+            messages::internalError(asyncResp->res);
+            return;
+        }
+        resetLicenseActivationStatus(asyncResp);
+    },
+        "com.ibm.License.Manager", "/com/ibm/license",
+        "org.freedesktop.DBus.Properties", "Set",
```
Comment (baemyung) [inline redfish-core/lib/license_service.hpp:258]: Consider to use `sdbusplus::asio::setProperty()`

File: redfish-core/lib/license_service.hpp [hunk 6]
```diff
@@ -0,0 +1,678 @@
...
+    std::string licenseString;
+    if (!json_util::readJsonAction(req, asyncResp->res, "LicenseString",
+                                   licenseString))
+    {
+        return;
+    }
+
+    if (licenseString.empty())
+    {
+        messages::propertyMissing(asyncResp->res, "LicenseString");
```
Comment (baemyung) [inline redfish-core/lib/license_service.hpp:322]: Should this function return here?
```
      messages::propertyMissing(asyncResp->res, "LicenseString");
+     return;
```

File: redfish-core/lib/license_service.hpp [hunk 7]
```diff
@@ -0,0 +1,678 @@
...
+                        std::get_if<std::string>(&(findStatus->second));
+                    if (status == nullptr)
+                    {
+                        messages::internalError(asyncResp->res);
+                        return;
+                    }
+                    // Ignore D-bus propertyChanged signal for
+                    // status value "Pending"
+                    if (*status == "com.ibm.License.LicenseManager."
+                                   "Status.Pending")
```
Comment (baemyung) [inline redfish-core/lib/license_service.hpp:392]: The new clang-format expects not to break a string
```
     if (*status == "com.ibm.License.LicenseManager.Status.Pending")
```

File: redfish-core/lib/license_service.hpp [hunk 8]
```diff
@@ -0,0 +1,678 @@
...
+            }
+
+            uint64_t expirationTime = 0;
+            const uint32_t* deviceNumPtr = nullptr;
+            const std::string* serialNumPtr = nullptr;
+            const std::string* licenseNamePtr = nullptr;
+            const std::string* licenseTypePtr = nullptr;
+            const std::string* authorizationTypePtr = nullptr;
+            bool available = false;
+            bool state = false;
```
Comment (baemyung) [inline redfish-core/lib/license_service.hpp:524]: Consider to use `sdbusplus::unpackPropertiesNoThrow()`
(or have another refactoring PR?)

File: redfish-core/lib/license_service.hpp [hunk 9]
```diff
@@ -0,0 +1,678 @@
...
+inline void requestRoutesLicenseEntry(App& app)
+{
+    BMCWEB_ROUTE(app, "/redfish/v1/LicenseService/Licenses/<str>/")
+        .privileges({{"Login"}})
+        .methods(boost::beast::http::verb::get)(
+            [](const crow::Request&,
+               const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
+               const std::string& param) {
+        getLicenseEntryById(asyncResp, param);
+    });
```
Comment (baemyung) [inline redfish-core/lib/license_service.hpp:676]: Can L672-L676 be simplified by doing?
1)
```
.methods(boost::beast::http::verb::get)(
    std::bind_front(getLicenseEntryById, std::ref(app)));
```
2) modify getLicenseEntryById() to 
```
inline void
    getLicenseEntryById(App& app, 
                                const std::shared_ptr<bmcweb::AsyncResp>& asyncResp,
                              const std::string& licenseEntryID)
{
    if (!redfish::setUpRedfishRoute(app, req, asyncResp))
    {
        return;
    }
    ....

}

```

File: redfish-core/lib/service_root.hpp [hunk 10]
```diff
@@ -258,7 +258,10 @@ inline void handleServiceRootGetImpl(
     asyncResp->res.jsonValue["TelemetryService"]["@odata.id"] =
         "/redfish/v1/TelemetryService";
     asyncResp->res.jsonValue["Cables"]["@odata.id"] = "/redfish/v1/Cables";
-
+#ifdef BMCWEB_ENABLE_REDFISH_LICENSE
+    asyncResp->res.jsonValue["LicenseService"] = {
+        {"@odata.id", "/redfish/v1/LicenseService"}};
```
Comment (baemyung) [inline redfish-core/lib/service_root.hpp:263]:
```
    asyncResp->res.jsonValue["LicenseService"]["@odata.id"] =  "/redfish/v1/LicenseService";
```

File: redfish-core/lib/license_service.hpp [hunk 11]
```diff
@@ -0,0 +1,678 @@
...
+            [](const crow::Request&,
+               const std::shared_ptr<bmcweb::AsyncResp>& asyncResp) {
+        asyncResp->res.jsonValue["@odata.id"] = "/redfish/v1/LicenseService/";
+        asyncResp->res.jsonValue["@odata.type"] =
+            "#LicenseService.v1_0_0.LicenseService";
+        asyncResp->res.jsonValue["Name"] = "License Service";
+        asyncResp->res.jsonValue["Id"] = "LicenseService";
+
+        asyncResp->res.jsonValue["Licenses"] = {
+            {"@odata.id", "/redfish/v1/LicenseService/Licenses"}};
```
Comment (baemyung) [inline redfish-core/lib/license_service.hpp:208]: According to https://github.com/openbmc/bmcweb/blob/master/COMMON_ERRORS.md#13-complete-replacement-of-the-response-object,

```
        asyncResp->res.jsonValue["Licenses"]["@odata.id"] = 
                    "/redfish/v1/LicenseService/Licenses";
```

File: redfish-core/lib/license_service.hpp (hunk 3 above)
Comment (gtmills) [reply redfish-core/lib/license_service.hpp:432]: If you want to rebase and use, I think it is better but I won't require it

File: redfish-core/lib/license_service.hpp (hunk 2 above)
Comment (gtmills) [reply redfish-core/lib/license_service.hpp:670]: If you want to rebase and use, I think it is better but I won't require it

File: redfish-core/include/redfish.hpp [hunk 12]
```diff
@@ -190,7 +191,7 @@ class RedfishService
         requestRoutesMemory(app);
 
         requestRoutesSystems(app);
-
+        requestRoutesSystemActionsOemExecutePanelFunction(app);
```
Comment (gtmills) [inline redfish-core/include/redfish.hpp:194]: this a merge conflict problem ?

File: redfish-core/lib/license_service.hpp [hunk 13]
```diff
@@ -0,0 +1,678 @@
...
+#include <array>
+#include <filesystem>
+#include <string_view>
+#include <variant>
+namespace redfish
+{
+namespace messages
+{
+static void addMessageToErrorJson(nlohmann::json& target,
+                                  const nlohmann::json& message)
```
[the remark this answers is not in the document]
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:21]: nope that makes future rebase difficult.That is not a modular design as well.  I was trying reduce the spread of the chnages  and restrict everything to single file. error_message.cpp has private functions. We need upstream changes to make some of them external reusable function. But that takes time. So for timebeing I just duplicated it.

File: redfish-core/lib/license_service.hpp [hunk 14]
```diff
@@ -0,0 +1,678 @@
...
+
+        asyncResp->res.jsonValue["Licenses"] = {
+            {"@odata.id", "/redfish/v1/LicenseService/Licenses"}};
+        asyncResp->res.jsonValue["Actions"] = {
+            {"#LicenseService.Install",
+             {{
+                 "target",
+                 "/redfish/v1/LicenseService/Actions/"
+                 "LicenseService.Install",
+             }}}};
```
[the remark this answers is not in the document]
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:215]: I don't see much value in following this recommendation. I would rather keep it as it is today.

File: redfish-core/lib/license_service.hpp (hunk 8 above)
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:524]: i would prefer not to make any changes here. I have seen same pattern used other places too. I don't want to make this a big refactoring work.

File: redfish-core/lib/license_service.hpp (hunk 7 above)
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:392]: did not get any format error

File: redfish-core/lib/license_service.hpp (hunk 11 above)
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:208]: I would prefer not to do this change for rebase PR

File: redfish-core/include/redfish.hpp (hunk 12 above)
Comment (abhilashraju) [reply redfish-core/include/redfish.hpp:194]: I am not sure how this happened

File: redfish-core/lib/license_service.hpp (hunk 2 above)
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:670]: I am not planning to do this change in this PR. Will take it up as separate refactoring line item

File: redfish-core/lib/license_service.hpp (hunk 3 above)
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:432]: I will not do it as part of this PR. Will add separate line item for it

File: redfish-core/lib/license_service.hpp [hunk 15]
```diff
@@ -0,0 +1,678 @@
...
+    },
+        "com.ibm.License.Manager", "/com/ibm/license",
+        "org.freedesktop.DBus.Properties", "Set",
+        "com.ibm.License.LicenseManager", "LicenseString",
+        std::variant<std::string>(licenseString));
+}
+inline void requestRoutesLicenseEntryCollection(App& app)
+{
+    BMCWEB_ROUTE(app, "/redfish/v1/LicenseService/Licenses")
+        .privileges({{"Login"}})
```
[the remark this answers is not in the document]
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:418]: Do it in separate PR

File: redfish-core/lib/license_service.hpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:215]: If you insists , will do it as another PR

File: redfish-core/lib/license_service.hpp [hunk 16]
```diff
@@ -0,0 +1,678 @@
...
+        for (const auto& object : resp)
+        {
+            std::string entryID = object.first.filename();
+            if (entryID.empty())
+            {
+                continue;
+            }
+            entriesArray.push_back({});
+            nlohmann::json& thisEntry = entriesArray.back();
+            thisEntry["@odata.id"] = "/redfish/v1/LicenseService/Licenses/" +
```
[the remark this answers is not in the document]
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:184]: Another PR

File: redfish-core/lib/license_service.hpp (hunk 9 above)
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:676]: why would that be better. Then we need to change the getLicenseEntryById parameters to accept const crow::Request& arg as well. I don't see bind_front will do any better to ignore argument. I prefer keep it same

File: redfish-core/include/redfish.hpp (hunk 12 above)
Comment (abhilashraju) [reply redfish-core/include/redfish.hpp:194]: is this becuase of this [rfrandse](https://github.com/rfrandse) [force-pushed](https://github.com/ibm-openbmc/bmcweb/compare/cf9657dae061c5afb8308f70534b46a711bcf325..bae1bf01096ea6fd4e5733d9dc2c9b79d8da9a6a) the 1110 branch from [cf9657d ](https://github.com/ibm-openbmc/bmcweb/commit/cf9657dae061c5afb8308f70534b46a711bcf325)to [bae1bf0 ](https://github.com/ibm-openbmc/bmcweb/commit/bae1bf01096ea6fd4e5733d9dc2c9b79d8da9a6a)
[8 hours ago](https://github.com/ibm-openbmc/bmcweb/pull/902#event-12497219864) ?

Comment (abhilashraju) [reply redfish-core/include/redfish.hpp:194]: I think I should create a new branch name different from 1110 as you suggested

**Comment by abhilashraju** [comment]
some merge conflict problem seen in this PR. So closing it . Will reopen new one

File: redfish-core/lib/license_service.hpp [hunk 17]
```diff
@@ -0,0 +1,678 @@
...
+#include <app.hpp>
+#include <error_messages.hpp>
+#include <registries/privilege_registry.hpp>
+#include <utils/json_utils.hpp>
+
+#include <array>
+#include <filesystem>
+#include <string_view>
+#include <variant>
+namespace redfish
```
[the remark this answers is not in the document]
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:16]: done

File: redfish-core/lib/license_service.hpp (hunk 5 above)
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:258]: will do it as part of another

File: redfish-core/lib/license_service.hpp (hunk 6 above)
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:322]: done

File: redfish-core/lib/license_service.hpp (hunk 7 above)
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:392]: done

File: redfish-core/lib/service_root.hpp (hunk 10 above)
Comment (abhilashraju) [reply redfish-core/lib/service_root.hpp:263]: done

File: redfish-core/lib/license_service.hpp (hunk 11 above)
Comment (abhilashraju) [reply redfish-core/lib/license_service.hpp:208]: done

**Pull request closed without merging**